File: petclinic-data/src/main/java/xyz/haff/petclinic/repositories/OwnerRepositoryCustomImpl.java
package xyz.haff.petclinic.repositories;

import lombok.RequiredArgsConstructor;
import org.springframework.r2dbc.core.DatabaseClient;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import xyz.haff.petclinic.models.Owner;
import xyz.haff.petclinic.models.User;

import java.util.UUID;

@Repository
@RequiredArgsConstructor
public class OwnerRepositoryCustomImpl implements OwnerRepositoryCustom<Owner> {
    private final DatabaseClient dbClient;
    private final UserRepository userRepository;

    @Override
    public <S extends Owner> Mono<S> save(S owner) {
        return userRepository.save(owner.getUser()).flatMap(user -> {
            if (owner.getVersion() == 0)
                return dbClient.sql("INSERT INTO owner (id, version, user, first_name, last_name) " +
                        "VALUES (:id, :version, :user, :firstName, :lastName)")
                        .bind("id", owner.getId())
                        .bind("version", 1)
                        .bind("user", user.getId())
                        .bind("firstName", owner.getFirstName())
                        .bind("lastName", owner.getLastName())
                        .then().then(Mono.just(owner))
                        ;
            else
                return Mono.error(RuntimeException::new);
        });
    }
}

File: petclinic-data/src/main/kotlin/xyz/haff/petclinic/repositories/OwnerRepositoryCustomImpl.kt
package xyz.haff.petclinic.repositories

import org.springframework.data.r2dbc.repository.Query
import org.springframework.r2dbc.core.DatabaseClient
import org.springframework.stereotype.Repository
import reactor.core.publisher.Mono
import xyz.haff.petclinic.models.Owner

@Repository
open class OwnerRepositoryCustomImpl(private val dbClient: DatabaseClient,
                                     private val userRepository: UserRepository): OwnerRepositoryCustom<Owner> {

    override fun <S : Owner?> save(owner: S): Mono<S> {
        return userRepository.save(owner!!.user).flatMap { (id) ->
            if (owner.version == 0) return@flatMap dbClient.sql(
                "INSERT INTO owner (id, version, user, first_name, last_name) " +
                        "VALUES (:id, :version, :user, :firstName, :lastName)"
            )
                .bind("id", owner.id)
                .bind("version", 1)
                .bind("user", id)
                .bind("firstName", owner.firstName)
                .bind("lastName", owner.lastName)
                .then()
                .then(Mono.just(owner))
            else return@flatMap Mono.error { RuntimeException() } // TODO
        }
    }
}